src/librbd/ImageWatcher.cc

```diff
@@ -576,6 +576,14 @@ void ImageWatcher<I>::schedule_request_lock(bool use_timer, int timer_delay) {
     return;
   }
 
+  if (is_blocklisted()) {
+    lderr(m_image_ctx.cct) << __func__ << ": "  << this
+                           << " blocklisted waiting for exclusive lock"
+                           << dendl;
+    m_image_ctx.exclusive_lock->handle_peer_notification(0);
+    return;
+  }
+
```

Comment:
I have not functional tested the following code,
```
diff --git a/src/librbd/ImageWatcher.cc b/src/librbd/ImageWatcher.cc
index dbbb3b5e112..38be5995d00 100644
--- a/src/librbd/ImageWatcher.cc
+++ b/src/librbd/ImageWatcher.cc
@@ -576,16 +576,7 @@ void ImageWatcher<I>::schedule_request_lock(bool use_timer, int timer_delay) {
     return;
   }
 
-  if (is_blocklisted()) {
-    lderr(m_image_ctx.cct) << __func__ << ": "  << this
-                           << " blocklisted waiting for exclusive lock"
-                           << dendl;
-    m_image_ctx.exclusive_lock->handle_peer_notification(0);
-    return;
-  }
-
-  std::shared_lock watch_locker{this->m_watch_lock};
-  if (this->is_registered(this->m_watch_lock)) {
+  if (is_registered()) {
     ldout(m_image_ctx.cct, 15) << this << " requesting exclusive lock" << dendl;
 
     auto ctx = new LambdaContext([this](int r) {
@@ -603,6 +594,11 @@ void ImageWatcher<I>::schedule_request_lock(bool use_timer, int timer_delay) {
     } else {
       m_task_finisher->queue(TASK_CODE_REQUEST_LOCK, ctx);
     }
+  } else if (is_blocklisted()) {
+    lderr(m_image_ctx.cct) << __func__ << ": "  << this
+                           << " blocklisted waiting for exclusive lock"
+                           << dendl;
+    return;
   }
 }
```
If you prefer the above changes, then I can test this too.

Reply:
Great! Then I will test this change by running the following test, https://tracker.ceph.com/issues/61607#note-2 for more than 6 hours.

Reply:
Please push the update to the PR first, just to make sure that we are on the same page and you would be testing the final revision.

Reply:
@idryomov the latest update looks good to you? The testing so far (> 2.5 hours) is looking good.

Reply:
LGTM!

Reply:
> The testing so far (> 2.5 hours) is looking good.

Just as before, please post a log snippet with an example of recovery.